src/commands/kv/bucket/sync.rs

```diff
@@ -19,11 +19,35 @@ pub fn sync(
     path: &Path,
     verbose: bool,
 ) -> Result<(), failure::Error> {
-    // First, upload all existing files in given directory
+    // First, upload all changed files in given local directory (aka replace files
+    // in Workers KV that are now stale).
+
+    // Get remote keys, which contain the hash of the file (value) as the suffix.
+    // Turn it into a HashSet. This will be used by upload() to figure out which
+    // files to exclude from upload (because their current version already exists in
+    // the Workers KV remote).
+    let remote_keys_iter = KeyList::new(target, user.clone(), namespace_id, None)?;
+    let mut remote_keys: HashSet<String> = HashSet::new();
+    for remote_key in remote_keys_iter {
+        match remote_key {
+            Ok(remote_key) => {
+                remote_keys.insert(remote_key.name);
+            }
+            Err(e) => failure::bail!(kv::format_error(e)),
+        }
+    }
+
     if verbose {
         message::info("Preparing to upload updated files...");
     }
-    upload_files(target, user.clone(), namespace_id, path, verbose)?;
+    upload_files(
+        target,
+        user.clone(),
+        namespace_id,
+        path,
+        Some(remote_keys.clone()),
```

Comment:
can this not take a reference?

Reply:
It can! doing a few edits to the function signatures so that they take more memory-efficient references rather than copy :)